Display MMU 3.x.x status messages in Navbar,

Kevman323 opened the pull request:
Hello, and thanks for your work on this plugin! I've been using it for a while now.

I've been doing a lot of research into how the MMU output works, so I decided to implement a lot of what I've learned here, to display more info in the navbar. 

I'm pretty new to python, and almost completely new to GitHub and Octoprint plugins, so sorry if I did something wrong. I've tried to replicate your code style, but I've changed things a lot in some areas. If you need to edit anything for any reason, go right ahead.

Other than that, I've thoroughly tested on my 3.0.0 MMU, including a small multicolor print, and I don't see any issues. I haven't tested on a lower MMU firmware, but in theory it should still work the same way it used to. Might be good to give it a quick test if possible though.

Here's a quick video I made showing the new features and fixes in action. https://www.youtube.com/watch?v=KW_aa21Af64

New Features:
- Improved MMU 3.x.x State detection
- Detection and display of "Preload to MMU", "Cutting", and "Ejecting" MMU States
- Display MMU progress messages in the navbar alongside the current state
- Display MMU error messages in the navbar alongside the current state

Fixed Bugs:
- Fix directional arrows being wrong in navbar for filament changes
- Fix previous filament showing up when doing a final unload
- Fix other strange navbar display issues
- Fixed a bunch of small bugs relating to MMU 3.x.x state detection edge cases

In-depth Changes:
- Added LOADING_MMU, CUTTING, and EJECTING states
  - I only coded detections for the MMU 3.x.x firmware, not sure if possible for older MMUs
- Completely redid MMU 3.x.x state checking
  - A general REGEX check in gcode_received_hook quickly returns most useless lines with just one check
  - A single REGEX search splits the MMU response into 4 convenient groups
  - All states except for ATTENTION and PAUSED_USER are found by looking at the MMU output
- Added an extra span to the navbar to display the extra message text for status messages and errors
  - The text is blank unless there's a progress or error message. It's also blank if the message is just F0 - Finished
- Removed MMU3Commands
  - I put a couple other lines in MMU3Codes instead. It's not quite right, but I thought having 4 different sections was kind of pointless
  - Added MMU3RequestCodes and MMU3ResponseCodes
  - Added a TON of documentation in Mmu.py about these codes
- Removed MmuKeys.ERROR
  - Added MmuKeys.RESPONSE which contains the MMU's response code: F, P, E, etc
  - Added MmuKeys.RESPONSE_DATA which contains the data from the response code. For example, the data after the E in Exxxx response codes
- Removed MmuKeys.LIVE_STATE
  - We can just use self.mmu[MmuKeys.STATE] to look up the current state before we change it
- Added MmuKeys.LAST_LINE
  - The entire last matched MMU line we care about is stored, and used for spamminess reduction in gcode_received_hook itself. Seems to work flawlessly with my tests
- Removed deduplication from _fire_event. All initial spamminess reduction is done in gcode_recieved_hook
  - Actual deduplication is all done in on_event. I've edited it to allow changes to response and responseData to fire MMU_CHANGED and update the navbar
- Added mmuProgress.js, which allows looking up the progress messages that the MMU sends while responding
- Filled out the remaining error messages in mmuErrors.js
  - As noted in the comment, it's possible for the Temperature and Electrical errors to have multiple cases at once, and produce a strange code. It's a complicated fix, so for now I just filled in the error codes for the single cases, which should be the most common
  - Removed the E from the error codes. Since the E response code and the responseData are split up now, it's easier to look at the single letter response code in prusammu.js, and just send the responseData itself to mmuErrors.js. mmuProgress.js works the same way
- Changed STATES.OK text from "Ready" to "Idle". "Ready" looked strange when an error occurred: "Ready -> ERROR"
- Various other changes to fix subtle bugs as I came across them

Kevman323 committed:
Add some missing error codes

Kevman323 committed:
New Features:
Improved MMU 3.0.0 State detection
Detection and display of "Preload to MMU", "Cutting", and "Ejecting" MMU States
Display MMU progress messages in the navbar alongside the current state
Display MMU error messages in the navbar alongside the current state

Fixed Bugs:
Fix directional arrows being wrong in navbar for filament changes
Fix previous filament showing up when doing a final unload
Fix other strange navbar display issues

In-depth Changes:
Added LOADING_MMU, CUTTING, and EJECTING states
- I only coded detections for the MMU 3.0 firmware, not sure if possible for older MMUs
Completely redid MMU 3.0 state checking
- A single REGEX search splits the MMU response into 4 convenient groups
- All states except for ATTENTION and PAUSED_USER are found by looking at the MMU output
Added an extra span to the navbar to display the extra message text for status messages and errors
- The text is blank unless there's a progress or error message. It's also blank if the message is just F0 - Finished
Removed MMU3Commands
- I put a couple other lines in MMU3Codes instead. It's not quite right, but I thought having 4 different sections was kind of pointless
- Added MMU3RequestCodes and MMU3ResponseCodes
- Added a TON of documentation in Mmu.py about these codes
Removed MmuKeys.ERROR
- Added MmuKeys.RESPONSE which contains the mmu's response code: F,P,E,etc
- Added MmuKeys.RESPONSE_DATA which contains the data from the response code. For example, the data after the E in Exxxx response codes
Removed MmuKeys.LIVE_STATE
- We can just use self.mmu[MmuKeys.STATE] to look up the current state before we change it
Added MmuKeys.LAST_LINE
- The entire last matched MMU line we care about is stored, and used for deduplication in gcode_received_hook itself. Seems to work flawlessly with my tests
- A general REGEX check in gcode_received_hook quickly returns most useless lines with just one check
Changed deduplication in _fire_event to work when the state is the same, but the response or responseData are different
- Removed a lot of deduplication code in on_event that seemed unecessary
Added mmuProgress.js, which allows looking up the progress messages that the MMU sends while responding
Filled out the remaining error messages in mmuErrors.js
- As noted in the comment, it's possible for the Temperature and Electrical errors to have multiple cases at once, and produce a strange code. It's a complicated fix, so for now I just filled in the error codes for the single cases, which should be the most common
- Removed the E from the error codes. Since the E response code and the responseData are split up now, it's easier to look at the single letter response code in prusammu.js, and just send the responseData itself to mmuErrors.js. mmuProgress.js works the same way
Changed STATES.OK text from "Ready" to "Idle". "Ready" looked strange when an error occurred: "Ready -> ERROR"
Various other changes to fix subtle bugs as I came across them

Kevman323 committed:
Update README.md

Kevman323 committed:
Undid most of the changes to _fire_event and on_event related to deduplication, so it works more like it did before my previous changes
Spamminess is now reduced in gcode_received_hook, and proper deduplication happens in on_event
_fire_event is pretty barebones now, similar to before the 3.0.0 changes

jukebox42 commented:
Wow, appreciate the contribution. I'll carve time to look at this improvement this weekend.

jukebox42 commented on octoprint_prusammu/__init__.py:
```diff
@@ -231,52 +235,82 @@ def gcode_received_hook(self, comm, line, *args, **kwargs):
+
+      # Since we rely on other lines to determine attention, if we're already at attention, and the line has an error, throw attention again and add the error code, just in case. This won't be spammed because of LAST_LINE
+      if (self.mmu[MmuKeys.STATE] == MmuStates.ATTENTION):
+        if (matchedCommand.group(3) == MMU3ResponseCodes.ERROR):
```
you can merge this if statement with the one above:
```
if (self.mmu[MmuKeys.STATE] == MmuStates.ATTENTION and matchedCommand.group(3) == MMU3ResponseCodes.ERROR):
```

Do we need to break out if we're already at attention to save some processing power? This function gets run a lot :)

Kevman323 replied:
Ah, the actual mistake here is that the return is indented too far. I'll fix it and make sure it doesn't break anything.

The idea behind this is that if we get a new MMU line, but the attention hasn't cleared, we want to stay at attention, BUT if the new MMU line has an error associated with it, we want to update that error code.

Overall this section shouldn't get run too much, due to the general regex check bouncing back the junk MMU lines. Then the LAST_LINE returns it immediately if the LAST_LINE is the same, so all the checks in the MMU responses section should only happen once per new MMU response that we are looking for.

Kevman323 replied:
It turns out that the code was working as it needed to, but not how I was trying to program it to work. I un-indented the return, and that broke it! After figuring it out, I've simplified as you've suggested, and updated the comment to better describe what that section does.

jukebox42 commented on octoprint_prusammu/__init__.py:
```diff
@@ -318,12 +352,8 @@ def gcode_sent_hook(self, comm, phase, cmd, cmd_type, gcode,
-        self.mmu[MmuKeys.LIVE_STATE] = payload[MmuKeys.STATE]
-        self._event_bus.fire(key, payload=payload)
-    else:
+      # Steps are taken in gcode_received_hook to reduce the spamminess of events. Proper deduplication happens in on_event
```
You need to reduce the indentation on these 3 lines by 2 should only be 2 instead of 4 from the def.

jukebox42 commented on octoprint_prusammu/__init__.py:
```diff
@@ -343,28 +373,33 @@ def on_event(self, event, payload=None):
-        # Remove error if we are no longer in attention
-        error="" if MmuKeys.ERROR not in payload or payload[MmuKeys.STATE] != MmuStates.ATTENTION else payload[MmuKeys.ERROR])
+        state=payload[MmuKeys.STATE],
+        lastLine=self.mmu[MmuKeys.LAST_LINE],
```
on_event is fired asynchronously using the event queue (fire pushes on the event queue). This may or may not align with the line the response is related to depending on when it's processed. If we're using this anything we should instead send it in the payload during the event fire

Kevman323 replied:
Similar to LIVE_STATE, LAST_LINE isn't used for anything down the way, so since we're updating self.mmu here, It's better to pass in the current LAST_LINE value than one that might be in the payload.

As a side note, I now see how LIVE_STATE worked for checks. I currently use several self.mmu[MmuKeys.STATE] checks in gcode_received_hook, and gcode_sent_hook already used self.mmu[MmuKeys.STATE] and self.mmu[MmuKeys.TOOL]. If I'm understanding this right, these checks could fail if the corresponding event to update self.mmu hasn't fired yet, right?

I could add back LIVE_STATE, but that doesn't fix the TOOL check in gcode_sent_hook. How do you feel about updating self.mmu in _fire_event instead of in on_event, so self.mmu is synchronous? We would then pass self.mmu as the payload to on_event, and don't need to update self.mmu there.

I understand that we want to avoid extra processing time as much as possible here. All the MMU3 response checks are deduped, so they only trigger _fire_event once per new MMU response that is different, and there's some other logic in there to reduce it even more than that sometimes. The other two checks I added are single lines that don't spam in the log, so should only fire once. All that remains is the MMU2 portion of the code. PAUSED_USER definitely spams, but with self.mmu being synchronous, the spamminess reduction in gcode_recieved_hook should prevent it from firing _fire_event more than once. The other MMU2 commands would need looked into to see if they are spammy.

jukebox42 replied:
Hm, I was thinking about this today. I think the right direction is to move the logic into _fire_event. There isn't much logic happening so I think we can afford it, and it'll prevent any race conditions. It's a better move than re-introducing LIVE_STATE, which in hindsight i should have done that in the first place. Pre 3.0.0 the commands came in so sparse that it wasn't a problem.

jukebox42 replied:
I'm also ok merging this as is and before I cut a new release, I can do that refactoring. Happy to let you handle it in this pr, but it didn't feel right asking you to undo my mistake lol.

Kevman323 replied:
I think I'll let you handle it. Moving the stuff into _fire_event means adding some extra deduplication, and possibly making MMU_CHANGED useless, so I'll let you decide how you want to go about with it. 
Thanks for the quick responses and review!

jukebox42 replied:
Sounds good! Leme merge it from here, I'll do the refactoring and can do the testing with that version. Appreciate the work! I'll be sure to credit you.

jukebox42 commented on octoprint_prusammu/static/mmuProgress.js:
```diff
@@ -0,0 +1,56 @@
+ * https://github.com/prusa3d/Prusa-Firmware/blob/MK3/Firmware/mmu2_progress_converter.cpp
+ */
+const MMU2MmuProgressStrings = {
+
```
nitpick: remove line

jukebox42 commented on octoprint_prusammu/static/prusammu.js:
```diff
@@ -172,22 +201,46 @@ $(() => {
       self._toolState = state;
-      self._error = error;
+      self._response = response;
+	  self._responseData = responseData;
```
nitpick: fix indentation

jukebox42 reviewed (commented):
Did an initial code review. Only minor things. Will run some print tests next.

Kevman323 committed:
Cleanup and fixes

Simplified and better explained a piece of code early on in the MMU3 response detection responsible for updating the error during the ATTENTION state
Deduplication for the lines that trigger ATTENTION on the MMU3. Some errors send both lines, or spam one
Better explanation of some MMU3 lines in Mmu.py
Minor formatting edits

Known Issue: If events sent to the event bus are delayed, the lines that check the current state may be checking an old state, breaking checks. Currently waiting on feedback for a fix for this.
